# Common Data Formats {#data_formats_bioc}

```{r setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble(cache = TRUE)
```

Chapter on common data formats to store large-scale data in R and Bioconductor


## SummarizedExperiment

Data from a high-throughput biological assay, such as single-cell RNA-sequencing (scRNA-seq), will often be summarised as a matrix of counts, where rows correspond to features and columns to samples[^1].
Within **Bioconductor**, the *SummarizedExperiment* class is the recommended container for such data, offering a rich interface that tightly links assay measurements to data on the features and the samples.

[^1]: Higher-dimensional arrays may be appropriate for some types of assays.

The *SummarizedExperiment* class is used to store rectangular arrays of experimental results (*assays*). Here, each *assay* is drawn as a matrix but higher-dimensional arrays are also supported.

`r knitr::include_graphics("https://raw.githubusercontent.com/PeteHaitch/BioC2020_DelayedArray_workshop/master/vignettes/images/SummarizedExperiment.svg")`

Traditionally, the assay data are stored in-memory as an ordinary *array* object[^2]. 
Storing the data in-memory becomes a real pain with the ever-growing size of 'omics datasets. 
It is now not uncommon to collect $10,000-100,000,000$ measurements on $100 - 1,000,000$ samples, which would occupy $10-1,000$ gigabytes (Gb) if stored in-memory as ordinary R arrays.

[^2]: In R, a *matrix* is just a 2-dimensional *array*

The DelayedArray framework offers a solution to this problem. 
By wrapping an array-like object (typically an on-disk object) in a *DelayedArray* object, it allows one to perform common array operations on the data without loading the object in memory. 
In order to reduce memory usage and optimize performance, operations on the object are either delayed or executed using a block processing mechanism.

We will learn more about the [DelayedArray framework](#da_overview). 


## Sparse data


## HDF5


## TileDB


## VCF


## GDS


