# (PART) Workflows {-}

# TENxBrainData {#scrna-tenxbraindata}

```{r unref-setup, echo=FALSE, results="asis"}
library(rebook)
chapterPreamble(cache=TRUE)
```

## Introduction

Here, we examine the TENxBrainData on ExperimentHub. 
This contains approximately 1.3 million cells of varying types such as oligodendrocytes, microglia and neurons.
Individual cells were isolated using the 10 Genomics protocol using a UMI-based protocol.
After sequencing, expression was quantified by counting the number of unique molecular identifiers (UMIs) mapped to each gene.

## Data loading

We start by obtaining the dense matrix and creating a HDF5Matrix object (a type of DelayedArray) 

```{r loading}
library(ExperimentHub)
hub <- ExperimentHub()

# Query ExperimentHub to find the relevant resource.
# This dataset is available in two formats: a 'dense matrix' format and a
# 'HDF5-based 10X Genomics' format. We'll use the 'dense matrix' version for 
# this workshop.
query(hub, "TENxBrainData")

# Load the relevant resource.
# This will download the data and may take a little while on the first run. 
# The result will be cached, however, so subsequent runs avoid re-downloading 
# the data.
fname <- hub[["EH1040"]]

# The structure of this HDF5 file can be seen using the h5ls() command
# from the rhdf5 package:
rhdf5::h5ls(fname)

# The 1.3 Million Brain Cell Dataset is represented by the "counts" group. 
# We point the HDF5Array() constructor to this group to create a HDF5Matrix 
# object (a type of DelayedArray) representing the dataset:
tenx <- HDF5Array(filepath = fname, name = "counts")

# We subset the data to the first 1000 rows
tenx_subset <- tenx[, 1:1000]
```

We will use the `tenx_subset` object in other chapters.

## Session Info {-}

```{r sessionInfo, echo=FALSE, results='asis'}
rebook::prettySessionInfo()
```